### PR TITLE
Added pagination support to get open PRs and comments API request methods

### PR DESCRIPTION
**Pull request by Cyborg-Squirrel**

### CODE REVIEW

**Comment by my-review-buddy** [comment]
Okay, here's a review of the pull request, focusing on potential bugs, style issues, and improvements. I'm assuming a focus on robustness, readability, and maintainability.

```python
diff --git a/github_api.py b/github_api.py
index 8b323ee..7307b6d 100644
--- a/github_api.py
+++ b/github_api.py
@@ -89,14 +89,18 @@ def __get_json_response_headers(self):
             "X-GitHub-Api-Version": "2022-11-28"
             }

-    def __do_json_api_post(self, url, request):
+    def __paginated_response_has_more_pages(self, headers) -> bool:
+        """Checks if the GitHub API response indicates there are more pages"""
+        link_header_present = 'Link' in headers.keys()
+        if link_header_present:
+            return '; rel="next"' in headers.get('Link')
+        return False
+
+    def __do_json_api_post_json(self, url, request):
         """POSTs a Github api request, returns the response json"""
         req = json.dumps(request)
         print(f"Request: {req}")
-        r = requests.post(url, headers=self.__get_json_response_headers(), data=req, timeout=5)
-        r.raise_for_status()
-        return r.json()
-
-    def __do_json_api_get(self, url) -> Any:
+
+    def __do_json_api_get_json(self, url) -> Any:
         """Does a Github api request, returns the response json"""
         r = requests.get(url, headers=self.__get_json_response_headers(), timeout=5)
         r.raise_for_status()
         return r.json()
-
-    def __do_json_api_get(self, url) -> requests.Response:
+
+    def __do_json_api_get(self, url) -> requests.Response:
         """Does a Github api request, returns the response json"""
         r = requests.get(url, headers=self.__get_json_response_headers(), timeout=5)
         r.raise_for_status()
         return r
+
+
     def __do_json_api_request_raw_response(self, url, headers):
         """Does a Github api request, returns the raw text"""
         r = requests.get(url, headers=headers, timeout=5)
         r.raise_for_status()
         return r.text

     def get_open_prs(self) -> list[GitHubPr]:
         """Gets all open pull requests for all configured repositories"""
         pr_list = []
         for repo in self.config.repo_list:
             open_prs_url = f"{self.__API_BASE}/repos/{repo.owner}/{repo.name}/pulls?state=open"
-            open_prs_for_repo = self.__do_json_api_get(open_prs_url)
-            if open_prs_for_repo is not None and len(open_prs_for_repo) > 0:
-                pr_list.extend(GitHubPr.schema().load(open_prs_for_repo, many=True))
+
             has_pages_remaining = True
             page = 0
             while has_pages_remaining:
                 page = page + 1
                 open_prs_url_page = f"{open_prs_url}&page={page}"
                 open_prs_for_repo_response = self.__do_json_api_get(open_prs_url_page)
                 has_pages_remaining = self.__paginated_response_has_more_pages(
                     open_prs_for_repo_response.headers)
                 pr_list.extend(GitHubPr.schema().load(open_prs_for_repo_response.json(), many=True))
         return pr_list

     def get_comments_for_pr(self, pr: GitHubPr) -> list[GitHubComment]:
         """Gets all comments posted on a specified pr"""
+
         comments = []
         has_pages_remaining = True
         page = 0
         while has_pages_remaining:
             page = page + 1
             comments_url_page = f"{pr.comments_url}?page={page}"
             comments_response = self.__do_json_api_get(comments_url_page)
             has_pages_remaining = self.__paginated_response_has_more_pages(
                 comments_response.headers)
             comments.extend(GitHubComment.schema().load(comments_response.json(), many=True))
         return comments

     def get_pr_diff(self, pr: GitHubPr) -> str:
         """Gets the diff for the pull request in raw form (not json)"""
         repo = pr.head.repo
         pr_files_url = f"{self.__API_BASE}/repos/{repo.owner.login}/{repo.name}"\
             f"/pulls/{pr.number}/files"
-        pr_changed_files = self.__do_json_api_get(pr_files_url)
+        pr_changed_files = self.__do_json_api_get_json(pr_files_url)
         return GitHubChangedFile.schema().load(pr_changed_files, many=True)

     def get_changed_file_whole_contents(self, file: GitHubChangedFile) -> str:
         """Gets a single file's contents from a changed file object"""
         raw_contents_url = f"{self.__API_BASE}/repos/{file.repo.owner.login}/{file.repo.name}"\
             f"/pulls/{file.pr.number}/files/{file.sha}"
-        r = self.__do_json_api_get(raw_contents_url)
+        r = self.__do_json_api_get_json(raw_contents_url)
         return r.text

     def get_upstream_file_whole_contents(self, pr: GitHubPr, file: GitHubChangedFile) -> str:
         """Gets a single file's contents from the base branch"""
         upstream_contents_url = f"{self.__API_BASE}/repos/{file.repo.owner.login}/{file.repo.name}"\
             f"/contents/{file.path}?ref={file.head.sha}"
-        r = self.__do_json_api_get(upstream_contents_url)
+        r = self.__do_json_api_get_json(upstream_contents_url)
         return r.text

     def post_comment(self, pr: GitHubPr, content: str):
         """Posts a comment to the specified pull request"""
         comments_url = pr.comments_url
-        self.__do_json_api_post(comments_url, {'body': content})
+        self.__do_json_api_post_json(comments_url, {'body': content})
 ```

**Review Breakdown:**

1.  **Naming Consistency & Duplication:**
    *   You're introducing `__do_json_api_post_json` and `__do_json_api_get_json`.  This creates a naming pattern that should be applied more consistently.  Consider renaming `__do_json_api_post` to `__do_json_api_post_json` to be fully consistent.  This avoids confusion about which method returns JSON vs. raw response.
2.  **Error Handling:**
    *   The `r.raise_for_status()` calls are good.  However, consider adding more specific error handling, especially if you encounter non-200 status codes. For example, you might want to log the error message and/or retry the request.
3.  **Docstrings:**
    *   The docstrings are generally good, but consider adding information about potential exceptions raised by each method.
4.  **Pagination Logic:**
    *   The pagination logic is well-implemented, but be aware of potential issues with very large repositories with thousands of open PRs or comments.  The GitHub API might have rate limits that could be reached. You could consider adding some form of rate limiting or backoff mechanism.
5.  **Response Type Consistency:**
    *   In the `get_pr_diff`, `get_changed_file_whole_contents` and `get_upstream_file_whole_contents` functions you are using `__do_json_api_get_json` after using `__do_json_api_get`. This is good that you did this, however, you should think of a more general pattern or a default to one approach.
6.  **Response Object:**
    *   The `__do_json_api_get` currently returns a `requests.Response` object. This isn't always necessary. If you only need the JSON data, you can directly return `r.json()`. However, returning the full `Response` object can be useful for debugging or for accessing headers or status codes. The pattern should be made consistent.

**Overall:**

This is a well-structured and generally well-implemented pull request. The addition of pagination support is a valuable improvement. The minor suggestions above aim to improve consistency, robustness, and maintainability.  Address the response object and response method naming issues for best practice.

**Comment by Cyborg-Squirrel** [comment]
@my-review-buddy please review

**Comment by my-review-buddy** [comment]
Okay, let's review this pull request. Overall, the addition of pagination support is great, as the GitHub API limits the number of results per page.  However, there are a few areas that need improvement, including style issues, potential bugs, and opportunities for better design.

Here's a breakdown of my feedback, organized by category (Bugs, Style, Improvements):

**I. Bugs & Potential Issues**

*   **`__do_json_api_get` returns `requests.Response` instead of JSON:** The method signature change in `__do_json_api_get` from `-> Any` to `-> requests.Response` is a good start, but the calling code still assumes it returns a JSON object. This is a potential source of errors, especially when `self.__paginated_response_has_more_pages` needs the response headers.  Make sure all calls to `__do_json_api_get` handle the `requests.Response` object correctly.
*   **Error Handling within Pagination Loops:**  The pagination loops (in `get_open_prs`, `get_comments_for_pr`, and `get_changed_files`) lack explicit error handling.  If a request within a loop fails (e.g., rate limit exceeded, network issue), the entire process crashes. You should add `try...except` blocks to gracefully handle these situations, potentially with retry logic or logging.
*   **`__paginated_response_has_more_pages` Header Check:**  The check `'; rel="next"' in headers.get('Link')` in `__paginated_response_has_more_pages` is brittle. It relies on the exact string `"; rel="next"` being present. GitHub could potentially change the format of the `Link` header.  It's better to use more robust parsing, potentially using a regular expression or a dedicated Link header parser.  For example:

    ```python
    import re

    def __paginated_response_has_more_pages(self, headers) -> bool:
        """Checks if the GitHub API response indicates there are more pages"""
        link_header_present = 'Link' in headers.keys()
        if link_header_present:
            link_header = headers.get('Link')
            if link_header:
                match = re.search(r'\brel="next"\b', link_header)
                return bool(match)
        return False
    ```

**II. Style Issues**

*   **Naming Consistency:**  The code mixes `__do_json_api_get` and `__do_json_api_post_json`. Choose one consistent naming scheme, either both with `_` (e.g. `_do_json_api_get`, `_do_json_api_post_json`) or both with `__` (e.g. `__do_json_api_get`, `__do_json_api_post_json`). The double underscore suggests these are truly "private" methods, so keeping that consistency would be good.
*   **Magic Numbers:** Using hardcoded page numbers (e.g., `page = page + 1`) in multiple places.  Consider extracting this logic into a helper function or a constant if it's used extensively.
*   **Unnecessary Repetition:** The pattern of `has_pages_remaining = True`, `page = 0`, `while has_pages_remaining:`, and incrementing `page` is repeated in several methods.  Extract this into a reusable pagination helper function.
*   **Missing Docstrings:** The `__do_json_api_post_json` function is missing a docstring, consistent with other helper functions.

**III. Improvements (Design & Readability)**

*   **Pagination Helper Function:**  Create a dedicated helper function to handle the pagination logic.  This will encapsulate the loop, the `has_pages_remaining` check, and potentially even the error handling.  This will significantly improve readability and reduce code duplication.

    ```python
    def _paginate(self, url_template, data_parser, max_retries=3):
        """
        Paginates through a GitHub API endpoint, parsing and returning the data.

        Args:
            url_template: A function that takes a page number and returns the URL for that page.
            data_parser:  A function that takes a requests.Response object and parses the response.
            max_retries: Number of times to retry a failed request.

        Returns:
            A list of parsed data.
        """
        data = []
        page = 1
        has_pages_remaining = True
        retries = 0
        while has_pages_remaining and retries < max_retries:
            try:
                url = url_template(page)
                response = self.__do_json_api_get(url)
                has_pages_remaining = self.__paginated_response_has_more_pages(response.headers)
                data.extend(data_parser(response))
                page += 1
            except requests.exceptions.RequestException as e:
                print(f"Error during pagination (page {page}): {e}")
                retries += 1
                if retries >= max_retries:
                   raise  # Re-raise the exception after max retries
                # Consider adding a time.sleep() here to avoid overwhelming the API

        return data
    ```

    Then in `get_open_prs`:

    ```python
    def get_open_prs(self) -> list[GitHubPr]:
        """Gets a list of open PRs for all repositories in the configuration."""
        pr_list = []
        for repo in self.config.repo_list:
            open_prs_url_template = lambda page: f"{self.__API_BASE}/repos/{repo.owner}/{repo.name}/pulls?state=open&page={page}"
            pr_list.extend(
                _paginate(
                    self,
                    open_prs_url_template,
                    lambda response: GitHubPr.schema().load(response.json(), many=True),
                )
            )
        return pr_list
    ```
*   **Dependency Injection (Optional):**  Consider injecting the `GitHubAPI` instance into its methods if you want to make it easier to test and mock the API calls.
*   **Type Hints:** Be consistent with type hints. If you're using `requests.Response` for internal methods, make sure to type hint them accordingly.
*   **Constants:** Define constants for common API endpoints or parameters (e.g., `OPEN_PULL_REQUESTS_ENDPOINT`).



This comprehensive review should help you refine this pull request and produce more robust and maintainable code.  Please let me know if you have any questions about any of these suggestions.

**Comment by my-review-buddy** [comment]
Okay, let's review these changes with an eye toward potential issues, style improvements, and general best practices.  I've structured the review into categories: Bugs, Style/Readability, Improvements, and Minor Suggestions.

**Overall Impression:**

The addition of pagination support is excellent.  The code is generally well-structured and attempts to handle responses correctly. However, there's some room for improvement regarding error handling, type hinting, and consistent naming.  The repeated use of `.extend` looks a bit clunky.

**1. Bugs & Potential Issues**

*   **`__paginated_response_has_more_pages` robustness:** The check `'; rel="next"' in headers.get('Link')` is brittle.  GitHub's Link header might change its format in the future. It would be better to use a more robust parsing mechanism (e.g., using a library to parse the `Link` header).  Consider a more lenient approach to account for slight variations in the header value.
*   **`__do_paginated_request` infinite loop:** If the GitHub API returns a broken or malformed Link header, or the Link is missing, the `while` loop in `__do_paginated_request` will run indefinitely. This needs to be addressed with a maximum iteration count.
*   **Error Handling in Paginated Requests:**  The code currently raises exceptions if the API call fails (due to rate limits, network issues, or other problems) during pagination.  However, these errors aren't handled gracefully. The loop should have a `try...except` block to catch `requests.exceptions.RequestException` (and potentially others) and either:
    *   Retry the request (with exponential backoff).
    *   Abort the pagination and log an error.
    *   Return a partial result and log the error.
*   **Mixed List and Single Object in `__do_paginated_request`:** The repeated usage of  `if json_obj is list: ... else: ...` within `__do_paginated_request` is a potential source of errors if the API's behavior changes unexpectedly. The API should consistently return a list, even if it's an empty one.  Consider a more robust check or, ideally, a guarantee from the API documentation.
*   **`__do_json_api_post_json` and `_do_json_api_post` confusion**: The existence of two POST methods with subtle differences (one using `json.dumps` and one not) creates confusion and potential errors. Choose one consistently. `__do_json_api_post_json` seems like the better choice given that the request body is always a JSON object.

**2. Style/Readability**

*   **Naming Consistency:** While the prefix `__` is used for private methods, it's good to maintain consistency. Ensure all internal helper methods follow the same naming convention.
*   **Typing:**  Strengthen type hints to be more specific. For example, in `__do_json_api_get`, the return type is `Any`, which is not helpful.  Ideally, provide a concrete type, or use `typing.TypeVar` if the type is unknown.  Also, for method that accept `GitHubPr` object, type hint the parameter.
*   **Docstrings**: Docstrings are present, but could be more descriptive for internal helper methods. Explain the assumptions and potential limitations.
*   **Magic Numbers**: `5` (timeout value) is a magic number.  Define a constant for it.

**3. Improvements**

*   **Centralized Pagination Logic:** The pagination logic is currently duplicated in multiple methods (`get_open_prs`, `get_comments_for_pr`, `get_changed_files`).  Consider extracting this logic into a single reusable function that takes the API endpoint, page parameter name, and request function as arguments.  This would reduce code duplication and improve maintainability.
*   **Rate Limit Handling:**  The GitHub API has rate limits.  The code doesn't explicitly handle rate limit errors (HTTP 403). Add logic to check the `X-RateLimit-Remaining` and `X-RateLimit-Reset` headers and pause execution if necessary. Consider implementing a retry mechanism with exponential backoff.
*   **Consider using `dataclasses_json.config`**: The `dataclasses_json` library offers more control over how dataclasses are serialized and deserialized.  Using `dataclasses_json.config` would allow you to customize the behavior of the library.
*   **Simplify `extend` calls**:  Instead of `pr_list.extend(GitHubPr.schema().load(open_prs_for_repo, many=True))` try `pr_list.extend([GitHubPr.schema().load(obj) for obj in open_prs_for_repo]` (this works if you know `open_prs_for_repo` is always a list. if not, use `isinstance`).

**4. Minor Suggestions**

*   Add a log message when pagination is stopped due to an error.
*   Consider using `logging.getLogger(__name__)` for consistent logging.
*   Use f-strings more consistently.
*   Add comments to clarify complex logic.

**Example Code (Illustrative)**

```python
import requests
from dataclasses_json import dataclass_json, config
from typing import Any, List

# CONSTANTS
API_TIMEOUT = 10  # seconds
MAX_PAGINATION_ITERATIONS = 100  # Prevent infinite loops

# ... (Existing code) ...


def __do_paginated_request(self, url: str, page_param: str) -> List[Any]:
    """
    Fetches paginated data from the GitHub API.
    
    Args:
        url: The base API endpoint URL.
        page_param: The query parameter used for pagination (e.g., '&page=').

    Returns:
        A list of objects fetched from the API.  May be truncated if errors occur.
    """
    response_list = []
    has_pages_remaining = True
    page = 0
    while has_pages_remaining:
        page += 1
        paginated_url = f"{url}{page_param}{page}"
        try:
            response = self.__do_json_api_get(paginated_url)
            has_pages_remaining = self.__paginated_response_has_more_pages(response.headers)
            response_list.extend(response.json())
            if page >= MAX_PAGINATION_ITERATIONS:
                logging.warning("Maximum pagination iterations reached.  Stopping pagination.")
                break
        except requests.exceptions.RequestException as e:
            logging.error(f"Error during pagination: {e}")
            break  # or retry with backoff

    return response_list
```

By addressing these points, you're improving the reliability, maintainability, and robustness of the pagination support and the overall GitHub API integration. Remember to test thoroughly, especially around error conditions and edge cases.
